packages/flutter/lib/src/scheduler/binding.dart

```diff
@@ -469,46 +469,108 @@ abstract class SchedulerBinding extends BindingBase {
     return new Duration(microseconds: (rawDurationSinceEpoch.inMicroseconds / timeDilation).round() + _epochStart.inMicroseconds);
   }
 
+  /// The time stamp for the frame currently being processed.
+  ///
+  /// This is only valid while [handleBeginFrame] is running, i.e. while a frame
+  /// is being produced.
+  // TODO(ianh): Replace this when fixing https://github.com/flutter/flutter/issues/5469
+  Duration get currentFrameTimeStamp {
+    assert(_currentFrameTimeStamp != null);
+    return _currentFrameTimeStamp;
+  }
+  Duration _currentFrameTimeStamp;
+
+  int _debugFrameNumber = 0;
+
   /// Called by the engine to produce a new frame.
   ///
   /// This function first calls all the callbacks registered by
-  /// [scheduleFrameCallback]/[addFrameCallback], then calls all the
-  /// callbacks registered by [addPersistentFrameCallback], which
-  /// typically drive the rendering pipeline, and finally calls the
-  /// callbacks registered by [addPostFrameCallback].
+  /// [scheduleFrameCallback]/[addFrameCallback], then calls all the callbacks
+  /// registered by [addPersistentFrameCallback], which typically drive the
+  /// rendering pipeline, and finally calls the callbacks registered by
+  /// [addPostFrameCallback].
+  ///
+  /// If the given time stamp is null, the time stamp from the last frame is
+  /// reused.
+  ///
+  /// To have a banner shown at the start of every frame in debug mode, set
+  /// [debugPrintBeginFrameBanner] to true. The banner will be printed to the
+  /// console using [debugPrint] and will contain the frame number (which
+  /// increments by one for each frame), and the time stamp of the frame. If the
+  /// given time stamp was null, then the string "warm-up frame" is shown
+  /// instead of the time stamp. This allows you to distinguish frames eagerly
+  /// pushed by the framework from those requested by the engine in response to
+  /// the vsync signal from the operating system.
+  ///
+  /// You can also show a banner at the end of every frame by setting
+  /// [debugPrintEndFrameBanner] to true. This allows you to distinguish log
+  /// statements printed during a frame from those printed between frames (e.g.
+  /// in response to events or timers).
   void handleBeginFrame(Duration rawTimeStamp) {
     Timeline.startSync('Frame');
     _firstRawTimeStampInEpoch ??= rawTimeStamp;
-    Duration timeStamp = _adjustForEpoch(rawTimeStamp);
+    _currentFrameTimeStamp = _adjustForEpoch(rawTimeStamp ?? _lastRawTimeStamp);
+    if (rawTimeStamp != null)
+      _lastRawTimeStamp = rawTimeStamp;
+
+    String debugBanner;
     assert(() {
-      if (debugPrintBeginFrameBanner)
-        debugPrint('━━━━━━━┫ Begin Frame ($timeStamp) ┣━━━━━━━');
+      _debugFrameNumber += 1;
+      if (debugPrintBeginFrameBanner || debugPrintEndFrameBanner) {
+        StringBuffer frameTimeStampDescription = new StringBuffer();
```

Comment:
Maybe a _debugPrintTimeStamp() or a _debugTimeStampToString() method would make it easier to see what's going on here in handleBeginFrame().